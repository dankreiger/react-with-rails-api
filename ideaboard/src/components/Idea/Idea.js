import React, { Component } from 'react';
import { bool, shape, string } from 'prop-types';
import { IdeaWrapper, IdeaBodyText } from './Idea.styles';
import { Col, Form, FormGroup, Input } from 'reactstrap';
import axios from 'axios';

class Idea extends Component {
  constructor(props) {
    super(props);
    this.state = {
      title: this.props.title,
      body: this.props.body,
      currentlyEditing: this.props.currentlyEditing
    };
  }

  handleInput = e => {
    this.setState({ [e.target.name]: e.target.value });
  };

  handleBlur = () => {
    axios
      .put(`http://localhost:3001/api/v1/ideas/${this.props.id}`, {
        idea: { title: this.state.title, body: this.state.body }
      })
      .then(response => {
        console.log(response);
        this.setState({currentlyEditing: !this.state.title.length || !this.state.body.length})
      })
      .catch(error => {
        console.log(error);
      });
  };
  render() {
    const { body, title, currentlyEditing } = this.state;
    return (
      <IdeaWrapper>
        {currentlyEditing ? (
          <Form onBlur={this.handleBlur}>
            <FormGroup row>
              <Col>
                <Input
                  type="text"
                  name="title"
                  placeholder="Title"
                  onChange={this.handleInput}
                />
              </Col>
            </FormGroup>
            <FormGroup row>
              <Col>
                <Input
                  type="textarea"
                  name="body"
                  placeholder="Idea"
                  onChange={this.handleInput}
                />
              </Col>
            </FormGroup>
          </Form>
        ) : (
          <>
            <p className="lead">{title}</p>
            <IdeaBodyText>{body}</IdeaBodyText>
          </>
        )}
      </IdeaWrapper>
    );
  }
}

Idea.defaultProps = {
  currentlyEditing: false
}
Idea.propTypes = {
  idea: shape({
    currentlyEditing: bool,
    title: string,
    body: string
  })
};

export default Idea;
