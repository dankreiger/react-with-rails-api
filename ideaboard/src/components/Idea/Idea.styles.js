import styled from 'styled-components';
import { Card } from 'reactstrap';

const fontSize = 16;
const lineHeight = 1.2;
const linesToShow = 3;
const fallbackHeight = fontSize * lineHeight * linesToShow;

export const IdeaWrapper = styled(Card)`
  display: flex;
  align-items: center;
  justify-content: center;
  margin-bottom: 20px;
  padding: 20px;
  height: 175px;
`;

export const IdeaTitleText = styled.p`
  white-space: nowrap;
  overflow: hidden;
  width: 100%;
  text-overflow: ellipsis;
`;

export const IdeaBodyText = styled.p`
  display: block; /* Fallback for non-webkit */
  display: -webkit-box;
  height: ${fallbackHeight}px; /* Fallback for non-webkit */
  margin-bottom: 0;
  font-size: ${fontSize}px;
  line-height: ${lineHeight};
  -webkit-line-clamp: ${linesToShow};
  -webkit-box-orient: vertical;
  overflow: hidden;
  text-overflow: ellipsis;
`;
