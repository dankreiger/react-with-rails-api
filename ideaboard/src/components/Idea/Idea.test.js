import React from 'react';
import { shallow } from 'enzyme';
import Idea from './Idea';

describe('Idea', () => {
  let ideaComponent;

  beforeEach(() => {
    ideaComponent = shallow(<Idea />);
  });

  describe('rendering', () => {
    test('renders as expected', () => {
      expect(ideaComponent.length).toBe(1);
      expect(ideaComponent).toMatchSnapshot();
    });
  });
});
