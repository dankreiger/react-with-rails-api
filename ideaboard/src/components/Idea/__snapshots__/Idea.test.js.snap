// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Idea rendering renders as expected 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <Idea
    currentlyEditing={false}
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "function",
    "props": Object {
      "children": <React.Fragment>
        <ForwardRef
          className="lead"
        />
        <ForwardRef />
      </React.Fragment>,
    },
    "ref": null,
    "rendered": Object {
      "instance": null,
      "key": undefined,
      "nodeType": "function",
      "props": Object {
        "children": Array [
          <ForwardRef
            className="lead"
          />,
          <ForwardRef />,
        ],
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "function",
          "props": Object {
            "children": undefined,
            "className": "lead",
          },
          "ref": null,
          "rendered": null,
          "type": Object {
            "$$typeof": Symbol(react.forward_ref),
            "attrs": Array [],
            "componentStyle": ComponentStyle {
              "componentId": "sc-bwzfXH",
              "isStatic": true,
              "rules": Array [
                "
  white-space: nowrap;
  overflow: hidden;
  width: 100%;
  text-overflow: ellipsis;
",
              ],
            },
            "displayName": "styled.p",
            "render": [Function],
            "styledComponentId": "sc-bwzfXH",
            "target": "p",
            "toString": [Function],
            "warnTooManyClasses": [Function],
            "withComponent": [Function],
          },
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "function",
          "props": Object {
            "children": undefined,
          },
          "ref": null,
          "rendered": null,
          "type": Object {
            "$$typeof": Symbol(react.forward_ref),
            "attrs": Array [],
            "componentStyle": ComponentStyle {
              "componentId": "sc-htpNat",
              "isStatic": true,
              "rules": Array [
                "
  display: block; /* Fallback for non-webkit */
  display: -webkit-box;
  height: ",
                "57.599999999999994",
                "px; /* Fallback for non-webkit */
  margin-bottom: 0;
  font-size: ",
                "16",
                "px;
  line-height: ",
                "1.2",
                ";
  -webkit-line-clamp: ",
                "3",
                ";
  -webkit-box-orient: vertical;
  overflow: hidden;
  text-overflow: ellipsis;
",
              ],
            },
            "displayName": "styled.p",
            "render": [Function],
            "styledComponentId": "sc-htpNat",
            "target": "p",
            "toString": [Function],
            "warnTooManyClasses": [Function],
            "withComponent": [Function],
          },
        },
      ],
      "type": Symbol(react.fragment),
    },
    "type": Object {
      "$$typeof": Symbol(react.forward_ref),
      "attrs": Array [],
      "componentStyle": ComponentStyle {
        "componentId": "sc-bdVaJa",
        "isStatic": true,
        "rules": Array [
          "
  display: flex;
  align-items: center;
  justify-content: center;
  margin-bottom: 20px;
  padding: 20px;
  height: 175px;
",
        ],
      },
      "displayName": "Styled(Card)",
      "render": [Function],
      "styledComponentId": "sc-bdVaJa",
      "target": [Function],
      "toString": [Function],
      "warnTooManyClasses": [Function],
      "withComponent": [Function],
      Symbol(Symbol.hasInstance): [Function],
    },
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "function",
      "props": Object {
        "children": <React.Fragment>
          <ForwardRef
            className="lead"
          />
          <ForwardRef />
        </React.Fragment>,
      },
      "ref": null,
      "rendered": Object {
        "instance": null,
        "key": undefined,
        "nodeType": "function",
        "props": Object {
          "children": Array [
            <ForwardRef
              className="lead"
            />,
            <ForwardRef />,
          ],
        },
        "ref": null,
        "rendered": Array [
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "function",
            "props": Object {
              "children": undefined,
              "className": "lead",
            },
            "ref": null,
            "rendered": null,
            "type": Object {
              "$$typeof": Symbol(react.forward_ref),
              "attrs": Array [],
              "componentStyle": ComponentStyle {
                "componentId": "sc-bwzfXH",
                "isStatic": true,
                "rules": Array [
                  "
  white-space: nowrap;
  overflow: hidden;
  width: 100%;
  text-overflow: ellipsis;
",
                ],
              },
              "displayName": "styled.p",
              "render": [Function],
              "styledComponentId": "sc-bwzfXH",
              "target": "p",
              "toString": [Function],
              "warnTooManyClasses": [Function],
              "withComponent": [Function],
            },
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "function",
            "props": Object {
              "children": undefined,
            },
            "ref": null,
            "rendered": null,
            "type": Object {
              "$$typeof": Symbol(react.forward_ref),
              "attrs": Array [],
              "componentStyle": ComponentStyle {
                "componentId": "sc-htpNat",
                "isStatic": true,
                "rules": Array [
                  "
  display: block; /* Fallback for non-webkit */
  display: -webkit-box;
  height: ",
                  "57.599999999999994",
                  "px; /* Fallback for non-webkit */
  margin-bottom: 0;
  font-size: ",
                  "16",
                  "px;
  line-height: ",
                  "1.2",
                  ";
  -webkit-line-clamp: ",
                  "3",
                  ";
  -webkit-box-orient: vertical;
  overflow: hidden;
  text-overflow: ellipsis;
",
                ],
              },
              "displayName": "styled.p",
              "render": [Function],
              "styledComponentId": "sc-htpNat",
              "target": "p",
              "toString": [Function],
              "warnTooManyClasses": [Function],
              "withComponent": [Function],
            },
          },
        ],
        "type": Symbol(react.fragment),
      },
      "type": Object {
        "$$typeof": Symbol(react.forward_ref),
        "attrs": Array [],
        "componentStyle": ComponentStyle {
          "componentId": "sc-bdVaJa",
          "isStatic": true,
          "rules": Array [
            "
  display: flex;
  align-items: center;
  justify-content: center;
  margin-bottom: 20px;
  padding: 20px;
  height: 175px;
",
          ],
        },
        "displayName": "Styled(Card)",
        "render": [Function],
        "styledComponentId": "sc-bdVaJa",
        "target": [Function],
        "toString": [Function],
        "warnTooManyClasses": [Function],
        "withComponent": [Function],
        Symbol(Symbol.hasInstance): [Function],
      },
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
}
`;
